src/db/mysql/models.rs

```diff
@@ -0,0 +1,352 @@
+use std::{self, ops::Deref};
+
+use diesel::{
+    insert_into,
+    mysql::MysqlConnection,
+    r2d2::{ConnectionManager, Pool, PooledConnection},
+    sql_types::{BigInt, Integer, Text},
+    update, Connection, ExpressionMethods, OptionalExtension,
+};
+use diesel::{sql_query, QueryDsl, RunQueryDsl};
+#[cfg(test)]
+use diesel_logger::LoggingConnection;
+use futures::future;
+
+use super::schema::{bso, collections, user_collections};
+#[cfg(test)]
+use super::test::TestTransactionCustomizer;
+use db::{
+    error::DbError, get_std_collection_id, get_std_collection_name, params, results,
+    util::ms_since_epoch, Db, DbFuture, Sorting,
+};
+use settings::Settings;
+
+embed_migrations!();
+
+pub type Result<T> = std::result::Result<T, DbError>;
+
+// The ttl to use for rows that are never supposed to expire (in seconds)
+pub const DEFAULT_BSO_TTL: u32 = 2100000000;
+
+no_arg_sql_function!(last_insert_id, Integer);
+
+/// Run the diesel embedded migrations
+///
+/// Mysql DDL statements implicitly commit which could disrupt MysqlPool's
+/// begin_test_transaction during tests. So this runs on its own separate conn.
+pub fn run_embedded_migrations(settings: &Settings) -> Result<()> {
+    let conn = MysqlConnection::establish(&settings.database_url).unwrap();
+    Ok(embedded_migrations::run(&conn)?)
+}
+
+pub struct MysqlDbPool {
+    pool: Pool<ConnectionManager<MysqlConnection>>,
+}
+
+// XXX: to become a db::DbPool trait
```

Comment:
How would this trait be used in practice?  It's not obvious to me whether the calling code really wants to know about pooling, or whether the connection pool could be a hidden implementation detail of a particular DB backend.

Reply:
Agreed, given how funky the Spanner backend is likely to be, pooling and such should prolly be hidden entirely behind the Db trait.

Reply:
Think of it as a basic factory for creating Dbs. I think regardless we require some kind of separation:

Db: directs calls, Send, !Sync
Factory: creates Dbs, Send + Sync